allennlp/training/trainer.py

```diff
@@ -319,12 +334,18 @@ def _train_epoch(self, epoch: int) -> Dict[str, float]:
 
             self.optimizer.zero_grad()
 
-            loss = self.batch_loss(batch_group, for_training=True)
+            # If num_accumulation_steps is 2 and num_gpus is 4, length of `batch_group`
+            # will be 8. This batch_group is split into 2 chunks each for a step, with each
+            # chunk consisting of 4 batches, 1 for each gpu.
+            batch_group_for_stepwise_accumulation = lazy_groups_of(iter(batch_group), len(self._cuda_devices))
+            for batch_for_step in batch_group_for_stepwise_accumulation:
```

Comment:
I am having a really tough time understanding what this code is doing here and why it's doing it, is there a way to clarify it / add comments

Reply:
Apologies if it is not succinct enough to understand. I've added few comments there. Hope it clarifies a bit. 

The basic idea is something like this. So far `batch_group` is being used to do a forward pass for both single and multi GPU cases. In the multi GPU case, the tensors in `batch_group` are aggregated to form a single batch. I just wanted to extend the same flow for gradient accumulation too. The inner `for` loop in `train_epoch` tries to call `batch_loss` for `num_steps_to_accumulate` times. In each iteration, we use a chunk of the original `batch_group` to call `batch_loss`. As usual, the length of the chunk that gets passed in each iteration would be equal to the number of GPUs configured for training.  Let me know if this makes sense.